### PR TITLE
Fuzz - Scenarios

### PR DESCRIPTION
**Pull request by julianmrodri**
Normal Scenario
* Setup StRSR as a registered token (so it is handled by echidna)
* Adds mutators for `distribute()` in Distributor and `manageToken()` in RevenueTrader.
* Adds tests for new mutators

Chaos Scenario
* Add functionality for Collateral mocks to default
* Adds all possible mutators (basket handler, asset registry, governance, etc)
* Adds randomness in trade settlement (auctions)
* Defines the Chaos scenario with collateral that can default
* Tests for Chaos Scenario

### CODE REVIEW

**Review by fiddlemath — COMMENTED** [review]
I didn't get through ChaosOps yet, but I took a look at everything else and made appropriate comments. All else looks good, and especially thank you for fixing the `between` bug everywhere!

File: contracts/fuzz/Utils.sol [hunk 1]
```diff
@@ -45,3 +45,13 @@ function between(
...
     return string(abi.encodePacked(a, b));
 }
+
+function getFirstChar(string memory originString) pure returns (string memory firstChar) {
+    bytes memory firstCharByte = new bytes(1);
+    firstCharByte[0] = bytes(originString)[0];
+    return string(firstCharByte);
+}
+
+function compareStr(string memory stringA, string memory stringB) pure returns (bool) {
```
[the remark this answers is not in the document]
Comment (julianmrodri) [reply contracts/fuzz/Utils.sol:55]: Done.

File: contracts/fuzz/CollateralMock.sol [hunk 2]
```diff
@@ -92,6 +112,48 @@ contract CollateralMock is OracleErrorMock, Collateral {
...
+        if (whenDefault <= block.timestamp) return;
+        CollateralStatus oldStatus = status();
+
+        // Check for hard default
+        uint192 referencePrice = refPerTok();
+
+        if (referencePrice < prevReferencePrice) {
+            whenDefault = block.timestamp;
+        } else {
+            uint192 p = (pricePerTarget() * targetPerRef()) / FIX_ONE;
```
[the remark this answers is not in the document]
Comment (julianmrodri) [reply contracts/fuzz/CollateralMock.sol:126]: Nice,if I do this then the `initialPeg` I store for comparison, should also only be the `targetPerRef()`.  Will implement.

File: contracts/fuzz/TradeMock.sol [hunk 3]
```diff
@@ -147,6 +186,57 @@ contract MarketMock is IMarketMock {
...
 
+    function calculateActualBuyAmt(IERC20 buy, uint256 buyAmt) internal returns (uint256) {
+        uint256 actualBuyAmt;
+
+        // Get next seed to calculate the actual buy amt
+        uint256 seed = getNextSeed();
+
+        uint256 maxTradeSlippage = getMaxTradeSlippage(buy);
+
+        if (mode == SettlingMode.Acceptable) {
```
[the remark this answers is not in the document]
Comment (julianmrodri) [reply contracts/fuzz/TradeMock.sol:197]: Isnt this going to get the "correct" value only 1 out of 5 times for the `Random` mode? Was that the intended behavior or we should be using `> 0`?

File: contracts/fuzz/TradeMock.sol [hunk 4]
```diff
@@ -147,6 +186,57 @@ contract MarketMock is IMarketMock {
...
+            actualBuyAmt = between(buyAmt, (buyAmt * (FIX_ONE + maxTradeSlippage)) / FIX_ONE, seed);
+        } else if (mode == SettlingMode.Random) {
+            // Allow to cause a violation in some cases
+            actualBuyAmt = between(
+                (buyAmt * (FIX_ONE - maxTradeSlippage)) / FIX_ONE,
+                GNOSIS_MAX_TOKENS - 1,
+                seed
+            );
```
[the remark this answers is not in the document]
Comment (julianmrodri) [reply contracts/fuzz/TradeMock.sol:205]: Yes, i wanted to allow it just a bit under the `minBuyAmount` to reduce the number of times this fails. But if we enforce that in another place we can fully randomize.

File: contracts/fuzz/TradeMock.sol [hunk 5]
```diff
@@ -147,6 +186,57 @@ contract MarketMock is IMarketMock {
...
+    function calculateActualBuyAmt(IERC20 buy, uint256 buyAmt) internal returns (uint256) {
+        uint256 actualBuyAmt;
+
+        // Get next seed to calculate the actual buy amt
+        uint256 seed = getNextSeed();
+
+        uint256 maxTradeSlippage = getMaxTradeSlippage(buy);
+
+        if (mode == SettlingMode.Acceptable) {
+            actualBuyAmt = between(buyAmt, (buyAmt * (FIX_ONE + maxTradeSlippage)) / FIX_ONE, seed);
```
[the remark this answers is not in the document]
Comment (julianmrodri) [reply contracts/fuzz/TradeMock.sol:198]: Done

File: contracts/fuzz/TradeMock.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (julianmrodri) [reply contracts/fuzz/TradeMock.sol]: converting to > 0 as we talked

File: contracts/fuzz/scenarios/ChaosOps.sol [hunk 6]
```diff
@@ -0,0 +1,909 @@
...
+    ) public {
+        // TODO: Review - More randomness for all params? Pricemodels, TradingRange, etc?
+        IERC20 erc20 = main.someToken(tokenID);
+        IAssetRegistry reg = main.assetRegistry();
+        if (!reg.isRegistered(erc20)) return;
+
+        IAsset asset = reg.toAsset(erc20);
+
+        // Switch type Asset -> Coll and viceversa one out of 100 times
+        bool switchType = (switchTypeSeed %= 100) <= 1;
```
Comment (fiddlemath) [inline contracts/fuzz/scenarios/ChaosOps.sol:373]: This will make switchType true 2 out of 100 times -- which is fine, but violates the stated intent.

Comment (fiddlemath) [reply contracts/fuzz/scenarios/ChaosOps.sol:373]: Also, using the `%=` operator here is slightly funky; why not just `%`?

**Review by fiddlemath — COMMENTED** [review]
OK, the whole ChaosOps scenario looks pretty solid to me, modulo the changes to asset registration that we already talked about. 👍

File: contracts/fuzz/FuzzP1.sol [hunk 7]
```diff
@@ -55,12 +70,45 @@ contract BasketHandlerP1Fuzz is BasketHandlerP1 {
...
 
 contract BackingManagerP1Fuzz is BackingManagerP1 {
     function _msgSender() internal view virtual override returns (address) {
         return IMainFuzz(address(main)).translateAddr(msg.sender);
     }
+
+    function invariantsHold() external view returns (bool) {
+        bool tradingDelayProp = (tradingDelay > MAX_TRADING_DELAY) ? false : true;
+        bool backingBufferProp = (backingBuffer > MAX_BACKING_BUFFER) ? false : true;
+        bool maxTradeSlippageProp = (maxTradeSlippage > MAX_TRADE_SLIPPAGE) ? false : true;
```
Comment (fiddlemath) [inline contracts/fuzz/FuzzP1.sol:108]:
```suggestion
        bool tradingDelayProp = tradingDelay <= MAX_TRADING_DELAY;
        bool backingBufferProp = backingBuffer <= MAX_BACKING_BUFFER;
        bool maxTradeSlippageProp = maxTradeSlippage <= MAX_TRADE_SLIPPAGE;
```

File: contracts/fuzz/FuzzP1.sol [hunk 8]
```diff
@@ -55,12 +70,45 @@ contract BasketHandlerP1Fuzz is BasketHandlerP1 {
         }
         return true;
     }
+
+    function invariantsHold() external view returns (bool) {
+        // if basket.erc20s is empty then disabled == true
+        bool disabledIfEmptyProp = (basket.erc20s.length == 0 && !basket.disabled) ? false : true;
```
Comment (fiddlemath) [inline contracts/fuzz/FuzzP1.sol:76]:
```suggestion
        bool disabledIfEmptyProp = basket.erc20s.length > 0 || basket.disabled;
```

File: contracts/fuzz/FuzzP1.sol [hunk 9]
```diff
@@ -101,12 +148,54 @@ contract BrokerP1Fuzz is BrokerP1 {
...
+    function invariantsHold() external view returns (bool) {
+        // (trades[addr] == true) iff this contract has created an ITrade clone at addr
+        bool tradesProp = true;
+        for (uint256 i = 0; i < tradeSet.length(); i++) {
+            if (!trades[tradeSet.at(i)]) tradesProp = false;
+        }
+
+        bool auctionLengthProp = (auctionLength == 0 || auctionLength > MAX_AUCTION_LENGTH)
+            ? false
+            : true;
```
Comment (fiddlemath) [inline contracts/fuzz/FuzzP1.sol:161]:
```suggestion
        bool auctionLengthProp = auctionLength > 0 && auctionLength <= MAX_AUCTION_LENGTH;
```

File: contracts/fuzz/FuzzP1.sol [hunk 10]
```diff
@@ -101,12 +148,54 @@ contract BrokerP1Fuzz is BrokerP1 {
...
         return IMainFuzz(address(main)).translateAddr(msg.sender);
     }
+
+    function invariantsHold() external view returns (bool) {
+        // ==== Invariants ====
+        // distribution is nonzero
+        RevenueTotals memory revTotals = totals();
+        bool distNotEmptyProp = (revTotals.rTokenTotal == 0 && revTotals.rsrTotal == 0)
+            ? false
+            : true;
```
Comment (fiddlemath) [inline contracts/fuzz/FuzzP1.sol:179]:
```suggestion
        bool distNotEmptyProp = !(revTotals.rTokenTotal == 0 && revTotals.rsrTotal == 0);
```

File: contracts/fuzz/FuzzP1.sol [hunk 11]
```diff
@@ -101,12 +148,54 @@ contract BrokerP1Fuzz is BrokerP1 {
...
+
+        // No invalid distributions to FURNACE and STRSR
+        bool noInvalidDistProp = (distribution[FURNACE].rsrDist > 0 ||
+            distribution[ST_RSR].rTokenDist > 0)
+            ? false
+            : true;
```
Comment (fiddlemath) [inline contracts/fuzz/FuzzP1.sol:185]:
```suggestion
        bool noInvalidDistProp = distribution[FURNACE].rsrDist == 0 &&
            distribution[ST_RSR].rTokenDist == 0;
```

File: contracts/fuzz/FuzzP1.sol [hunk 12]
```diff
@@ -119,6 +208,11 @@ contract RevenueTraderP1Fuzz is RevenueTraderP1 {
     function _msgSender() internal view virtual override returns (address) {
         return IMainFuzz(address(main)).translateAddr(msg.sender);
     }
+
+    function invariantsHold() external view returns (bool) {
+        bool maxTradeSlippageProp = (maxTradeSlippage > MAX_TRADE_SLIPPAGE) ? false : true;
```
Comment (fiddlemath) [inline contracts/fuzz/FuzzP1.sol:213]:
```suggestion
        bool maxTradeSlippageProp = maxTradeSlippage <= MAX_TRADE_SLIPPAGE;
```

File: contracts/fuzz/FuzzP1.sol [hunk 13]
```diff
@@ -175,7 +269,17 @@ contract StRSRP1Fuzz is StRSRP1 {
         bool stakesProp = totalStakes == 0 ? stakeRSR == 0 && stakeRate == FIX_ONE : stakeRSR > 0;
         bool draftsProp = totalDrafts == 0 ? draftRSR == 0 && draftRate == FIX_ONE : draftRSR > 0;
 
-        return stakesProp && draftsProp;
+        bool totalStakesCovered = true;
+        if (totalStakes > 0) {
+            if (stakeRSR * stakeRate < totalStakes * 1e18) totalStakesCovered = false;
+        }
```
Comment (fiddlemath) [inline contracts/fuzz/FuzzP1.sol:275]:
```suggestion
        bool totalStakesCovered = stakeRSR * stakeRate >= totalStakes * 1e18;
```

File: contracts/fuzz/FuzzP1.sol [hunk 14]
```diff
@@ -175,7 +269,17 @@ contract StRSRP1Fuzz is StRSRP1 {
...
+
+        bool totalDraftsCovered = true;
+        if (totalDrafts > 0) {
+            if (draftRSR * draftRate < totalDrafts * 1e18) totalDraftsCovered = false;
+        }
```
Comment (fiddlemath) [inline contracts/fuzz/FuzzP1.sol:280]:
```suggestion
        bool totalDraftsCovered = draftRSR * draftRate >= totalDrafts * 1e18;
```

**Review by fiddlemath — CHANGES_REQUESTED** [review]
I've got some style qualms with the way the invariant implementations, but otherwise the sense of all of this code seems good.

File: contracts/fuzz/FuzzP1.sol (hunk 7 above)
Comment (julianmrodri) [reply contracts/fuzz/FuzzP1.sol:108]: Done

File: contracts/fuzz/FuzzP1.sol (hunk 8 above)
Comment (julianmrodri) [reply contracts/fuzz/FuzzP1.sol:76]: Done

File: contracts/fuzz/FuzzP1.sol (hunk 9 above)
Comment (julianmrodri) [reply contracts/fuzz/FuzzP1.sol:161]: Done

File: contracts/fuzz/FuzzP1.sol (hunk 10 above)
Comment (julianmrodri) [reply contracts/fuzz/FuzzP1.sol:179]: Done

File: contracts/fuzz/FuzzP1.sol (hunk 11 above)
Comment (julianmrodri) [reply contracts/fuzz/FuzzP1.sol:185]: Done

File: contracts/fuzz/FuzzP1.sol (hunk 12 above)
Comment (julianmrodri) [reply contracts/fuzz/FuzzP1.sol:213]: Done

File: contracts/fuzz/FuzzP1.sol (hunk 13 above)
Comment (julianmrodri) [reply contracts/fuzz/FuzzP1.sol:275]: Done

File: contracts/fuzz/FuzzP1.sol (hunk 14 above)
Comment (julianmrodri) [reply contracts/fuzz/FuzzP1.sol:280]: Done

**Review by fiddlemath — APPROVED** [review]

**Comment by julianmrodri** [comment]
Merging to `fuzz` based on offline conversations.

**Pull request merged**